### PR TITLE
Server REST API

### PR DESCRIPTION
**Pull request by timgerstel**
REST API which enables administrators to query the status of ZSS server.

/server/agent -> list routes
/server/agent/config -> display server config in the form of zluxserver.json with respect to CLI args.
/server/agent/log -> display server log file
/server/agent/logLevels -> display logger verbosity levels
/server/agent/environment -> display information about the environment in which the ZSS server is running.

Do not merge without: 
https://github.com/zowe/zowe-common-c/pull/81

### CODE REVIEW

File: c/restService.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    if (!strcmp(request->method, methodGET)) {
+      char *l1 = stringListPrint(request->parsedFile, 2, 1, "/", 0);
+      if(!strcmp(l1, "")){
+        respondWithServerRoutes(response);
+      }
+      else if (!strcmp(l1, "config")){
+        respondWithServerConfig(response);
+      }
+      else if (!strcmp(l1, "log")) {
+        if(strcmp(getenv("ZSS_LOG_FILE"), "")){
```
Comment (ifakhrutdinov) [inline c/restService.c]: ```getenv``` calls may return NULL pointers. Please check for NULL before using the results of ```getenv```.

File: c/restService.c [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+static int serveStatus(HttpService *service, HttpResponse *response) {
+  HttpRequest *request = response->request;
+  JsonObject *dataserviceAuth = jsonObjectGetObject(serverConfig, "dataserviceAuthentication");
+  int rbacParm = jsonObjectGetBoolean(dataserviceAuth, "rbac");
+  if(!rbacParm){
+     respondWithError(response, HTTP_STATUS_UNAUTHORIZED, "Unauthorized - RBAC is disabled.  Enable in zluxserver.json");
+  } else {
+    if (!strcmp(request->method, methodGET)) {
+      char *l1 = stringListPrint(request->parsedFile, 2, 1, "/", 0);
```
Comment (ifakhrutdinov) [inline c/restService.c]: Can ```l1``` be NULL?

File: c/restService.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  jsonStart(out);
+  jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = strdup(*environ);
```
Comment (ifakhrutdinov) [inline c/restService.c]: According to the doc, ```strdup``` may return NULL.

File: c/restService.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  int i = 1;
+  for (; env_var; i++) {
+    char *var_name = strtok(env_var, "=");
+    char *var_value = strtok(NULL, "=");
+    jsonAddString(out, var_name, var_value);
+    env_var = strdup(*(environ+i));
```
Comment (ifakhrutdinov) [inline c/restService.c]: What if ```*(environ+i) == NULL```, shouldn't that be checked for NULL in the for loop instead of ```env_var```?

File: c/restService.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  jsonAddString(out, "agentVersion", productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = strdup(*environ);
+  int i = 1;
+  for (; env_var; i++) {
+    char *var_name = strtok(env_var, "=");
```
Comment (ifakhrutdinov) [inline c/restService.c]: What if there is no ```"="``` in the string?

File: c/restService.c [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+int installServerStatusService(HttpServer *server, JsonObject *serverSettings, char* productVer)
+{
+  HttpService *httpService = makeGeneratedService("REST_Service", "/server/agent/**");
+  httpService->authType = SERVICE_AUTH_NATIVE_WITH_SESSION_TOKEN;
+  httpService->serviceFunction = &serveStatus;
+  httpService->runInSubtask = TRUE;
+  httpService->doImpersonation = TRUE;
+  registerHttpService(server, httpService);
+  serverConfig = serverSettings;
+  memcpy(productVersion, productVer, 40);
```
[the remark this answers is not in the document]
Comment (timgerstel) [reply c/restService.c]: I should verify that it is null terminated but this line is how I know it will be equal to 40 bytes: https://github.com/zowe/zss/blob/master/c/zss.c#L84

File: c/serverStatusService.c [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+#include "restService.h"
+#pragma linkage (EXSMFI,OS)
+
+#ifdef __ZOWE_OS_ZOS
+static int serveStatus(HttpService *service, HttpResponse *response);
+static JsonObject* serverConfig = NULL;
+char productVersion[40];
+extern char **environ;
+
+typedef int EXSMFI();
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: the correct syntax for a function with no arguments is` EXSMFI(void)`. `EXSMFI()` means "a function which you can all with any number of arguments, all of them will be cast to int"

File: c/serverStatusService.c [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+               &subtype,
+               buffer,
+               &bufferlen,
+               &cpu_util,
+               &demand_paging,
+               &options,
+               &mvs_srm,
+               &zaap_util,
+               &ziip_util);
+  sprintf(pid, "%d", getpid());
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: I would instead to 
`snprintf(pid, sizeof (pid), "%d", getpid());`
etc

File: c/serverStatusService.c [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+void respondWithServerEnvironment(HttpResponse *response){
+  /*Information about parameters for smc_func: https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.erbb700/smfp.htm#smfp*/
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  struct utsname unameRet;
+  uname(&unameRet);
+  char pid[64];
+  char ppid[64];
+  char dp[64];
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: It's unclear from the context what `dp`, `mvs_u`, etc, are. Can you please add a little more descriptive names?

File: c/serverStatusService.c [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+  char cpu_u[64];
+  char mvs_u[64];
+  char ziip_u[64];
+  char zaap_u[64];
+  char* buffer;
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: We ignore the variable after the call. Is that intentional?

File: c/serverStatusService.c [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+  char *env_var = strdup(*environ);
+  int i = 1;
+  for (; env_var; i++) {
+    char *var_name = strtok(env_var, "=");
+    char *var_value = strtok(NULL, "=");
+    jsonAddString(out, var_name, var_value);
+    env_var = strdup(*(environ+i));
+  }
+  jsonEndObject(out);
+  jsonAddString(out, "demandPagingRate", dp);
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: Why don't we do this here instead:
```
chat smfVarBuffer[64];

snprintf(smfVarBuffer, sizeof (smfVarBuffer), "%d", demand_paging);
jsonAddString(out, "demandPagingRate", dp); 

snprintf(smfVarBuffer, sizeof (smfVarBuffer), "%d", cpu_util);
jsonAddString(out, "stdCP_CPU_Util", dp); 
etc...
```

File: c/serverStatusService.c [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+  char mvs_u[64];
+  char ziip_u[64];
+  char zaap_u[64];
+  char* buffer;
+  int rc = 0;
+  int reqtype = 0x00000005; //fullword. request type
+  int rectype = 0x0000004F; //SMF record type, only type 79 is supported
+  int subtype = 0x00000009; //SMF record subtype
+  int bufferlen = 247488; /* length of SMF record buffer */
+  int cpu_util = 0x00000000;
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: naming: cpuUtil, demandPaging, etc.

File: c/serverStatusService.c [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+  jsonEndObject(out);
+  jsonAddString(out, "demandPagingRate", dp); 
+  jsonAddString(out, "stdCP_CPU_Util", cpu_u);
+  jsonAddString(out, "stdCP_MVS_SRM_CPU_Util", mvs_u);
+  jsonAddString(out, "ZAAP_CPU_Util", zaap_u);
+  jsonAddString(out, "ZIIP_CPU_Util", ziip_u);
+  jsonAddString(out, "PID", pid);
+  jsonAddString(out, "PPID", ppid);
+  jsonEnd(out);
+  safeFree(env_var, sizeof(*environ)+1);
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: `*evniron` is just a pointer, so `sizeof(*environ)` is not the length of the string (although we don't care about that in LE, this is still misleading)

File: c/serverStatusService.c [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        respondWithLogLevels(response);
+      }
+      else if (!strcmp(l1, "environment")) {
+        respondWithServerEnvironment(response);
+      }
+      else {
+        respondWithJsonError(response, "Invalid path", 400, "Bad Request");
+      }
+    }
+    else{
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: formatting: please remove the newline before `else`

File: c/serverStatusService.c [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+  jsonEnd(out);
+  safeFree(env_var, sizeof(*environ)+1);
+  finishResponse(response);
+}
+
+static int serveStatus(HttpService *service, HttpResponse *response) {
+  HttpRequest *request = response->request;
+  JsonObject *dataserviceAuth = jsonObjectGetObject(serverConfig, "dataserviceAuthentication");
+  int rbacParm = jsonObjectGetBoolean(dataserviceAuth, "rbac");
+  if(!rbacParm){
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: The name of the variable and the config JSON field is confusing. How is this related to RBAC?

File: c/serverStatusService.c [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+  int cpu_util = 0x00000000;
+  int demand_paging = 0x00000000;
+  int options = 0x00000000;
+  int mvs_srm = 0x00000000;
+  int zaap_util = 0x00000000;
+  int ziip_util = 0x00000000;
+  buffer = (char *)safeMalloc(bufferlen, "buffer");
+  memset(buffer, 0, bufferlen);
+  smf_func = (EXSMFI *)fetch("ERBSMFI");
+  rc = (*smf_func)(&reqtype,
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: can `rc` indicate an error?

File: c/serverStatusService.c [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+  jsonStart(out);
+  jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = strdup(*environ);
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: Like Irek said above, null check

File: c/serverStatusService.c (hunk 15 above)
Comment (timgerstel) [reply c/serverStatusService.c]: The member that states whether RBAC is enabled is under the dataserviceAuthentication member.
```
dataserviceAuthentication{
  "defaultAuth...": "...",
  "rbac": true
  ....
}
```

**Comment by timgerstel** [comment]
@ifakhrutdinov @fkovinAtRocket Thank you guys very much for the verbose feedback, I appreciate you guys taking time to review this for me.  I have implemented all of the requested changes, if one or both of you wouldn't mind taking a look at the changes I would be very grateful!

File: c/restService.c (hunk 5 above)
Comment (timgerstel) [reply c/restService.c]: I do not believe this is a possible scenario for environment variables.  There must be at a minimum one "=" unless I am mistaken

File: c/restService.c (hunk 4 above)
Comment (timgerstel) [reply c/restService.c]: Doesn't the end condition "env_var;" mean "loop until env_var is null"?  In which case, when *(environ+i) == NULL, env_var will also be null and thus exit the loop.

File: c/restService.c (hunk 2 above)
Comment (timgerstel) [reply c/restService.c]: I do not believe stringListPrint() can return null.

File: c/zss.c [hunk 18]
```diff
@@ -817,7 +818,7 @@ static int validateFilePermissions(const char *filePath) {
   } else {
     return validateConfigPermissionsInner(filePath);
   }
-}
+}
```
Comment (ifakhrutdinov) [inline c/zss.c]: Please don't commit trailing spaces.

File: c/serverStatusService.c [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+      break;
+    }
+    char *name = safeMalloc(strlen(env_var)-strlen(len) + 1, "env_var name");
+    snprintf(name, sizeof(char)*(strlen(env_var)-strlen(len) + 1), "%s", env_var);
+    jsonAddString(out, name, len+1);
+    safeFree(name, strlen(name));
+    env_var = *(environ+i);
+  }
+  jsonEndObject(out);
+  snprintf(smfVarBuffer, sizeof(smfVarBuffer), "%d%%", demandPaging);
```
Comment (ifakhrutdinov) [inline c/serverStatusService.c]: Why do we need to convert integers to strings? Client of this service now have to parse these values (and take into account the fact that some values have a percent sign).

Can we do ```jsonAddInt```?

File: c/serverStatusService.c [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+               buffer,
+               &bufferlen,
+               &cpuUtil,
+               &demandPaging,
+               &options,
+               &mvsSrm,
+               &zaapUtil,
+               &ziipUtil);
+  if(rc > 0){
+    respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Unable to fetch from RMF data interface service");
```
Comment (fkovinAtRocket) [inline c/serverStatusService.c]: and then free the buffer and return, right?

File: c/serverStatusService.c (hunk 15 above)
Comment (fkovinAtRocket) [reply c/serverStatusService.c]: It's just not obvious how this is related to the rest of the code. Why would we not send someone a status response if rbac is disabled in the config?

File: c/serverStatusService.c (hunk 20 above)
Comment (timgerstel) [reply c/serverStatusService.c]: I free the buffer above line 156 now, its never used after and does not contain (human) readable data.  respondWithError returns the response.

**Comment by timgerstel** [comment]
> 
> 
> @timgerstel Please see my comments. I haven't finished reviewing yet, will look at the code more later.

Thank you again, I have implemented your requested changes.

File: c/restService.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (1000TurquoisePogs) [reply c/restService.c]: This can be best solved by using a `#DEFINE` to state that both are 40.

File: c/serverStatusService.c (hunk 15 above)
Comment (1000TurquoisePogs) [reply c/serverStatusService.c]: We need to send a response when RBAC is disabled, but it should be an HTTP 400 bad request. This service is conditional on RBAC being enabled, because it is a sensitive URL.

**Comment by 1000TurquoisePogs** [comment]
Also, dont forget DCO as it is challenging to fix the longer you wait.

File: c/serverStatusService.c (hunk 20 above)
Comment (fkovinAtRocket) [reply c/serverStatusService.c]: but we still need to return from the function now

File: c/serverStatusService.c (hunk 15 above)
Comment (fkovinAtRocket) [reply c/serverStatusService.c]: That would be a comment I was looking for :-)
